Add context-menu item for generating QR code for the current page (#321)

mahmoudhusam opened the pull request:
This PR adds a context-menu item to generate a QR code for the current page URL. The functionality is implemented in the ContextMenu.js module, and a new message string was added to messages.json for localization.
Resolves issue #321.

rugk commented on src/_locales/en/messages.json:
```diff
@@ -186,6 +186,14 @@
     "description": "The context menu entry shown for generating QR codes from a selected link with an access key."
   },
+  "contextMenuItemConvertPageURL": {
+    "message": "Generate QR Code for this Page",
+    "description": "Context menu item title for generating a QR code for the current page URL."
+  },
+  "contextMenuItemConvertPageURLAccessKey": {
+    "message": "&Generate QR Code for this Page",
```
```suggestion
    "message": "Generate QR code for this page",
    "description": "Context menu item title for generating a QR code for the current page URL."
  },
  "contextMenuItemConvertPageURLAccessKey": {
    "message": "&Generate QR code for this page",
```

AFAIK this is the capitalisation style we use in the translations for now, everywheere. Feel free to argue against this especially if you are a native speaker, but then changing this would need to be tackled in a new issue/PR and be consistent everywhere.

mahmoudhusam replied:
Thank you for the feedback! I've committed the suggested change.

rugk commented on src/background/modules/ContextMenu.js:
```diff
@@ -4,6 +4,7 @@
 const CONVERT_TEXT_SELECTION = "qr-convert-text-selection";
 const CONVERT_LINK_TEXT_SELECTION = "qr-convert-link-text-selection";
 const OPEN_OPTIONS = "qr-open-options";
+const CONVERT_PAGE_URL = "qr-convert-page-url";
```
Hu? This is fine.

rugk reviewed (changes requested):
Hi @mahmoudhusam,
first of all, thanks for your first contribution to this project! :tada: :+1: 🏅
I hope you'll like this project and enjoy hacking on it… :smiley:

It generally looks good, but I have some pointers to change:
* First of all, would you have a screenshot or even screencast to show (how) it works?
* Secondly, I see you may need to add removing that menu entry to the `menuShown` function like the others? I am not sure why, but maybe because it is re-rendered… and may not be shown every time? I am unsure, and unfortunately the JSDoc is also not helpful there (blame on me, but feel free to improve it)
* Generally, as explained in the issue before (https://github.com/rugk/offline-qr-code/issues/321#issuecomment-2158888590) the feature needs to be configurable. That is as far as I see still missing. See https://github.com/rugk/offline-qr-code/issues/157#issuecomment-508475513 and FYI for some guides/pointers on how to introduce such a setting.

mahmoudhusam commented:
Hi @rugk,

Thank you so much for your guidance and for helping me through my first contribution to open-source! I've learned a lot from this project, and your feedback has been incredibly valuable.

I've implemented the requested changes:

Added a screencast demonstrating the new feature in action (attached below).
Please let me know if there's anything else you'd like me to adjust. I'm more than happy to make further improvements!

Thanks again for your support.

![contextMenuItemConvertPageURL](https://github.com/user-attachments/assets/ed1eae4c-6cf8-401b-9189-dce427ea45d0)
![contextMenuEnabled](https://github.com/user-attachments/assets/ea9ada99-7812-4bc4-b344-018b982f1148)

rugk commented on src/options/modules/CustomOptionTriggers.js:
```diff
@@ -72,6 +72,37 @@ function applyPopupIconColor(optionValue) {
+ */
+document.getElementById('contextMenuEnabled').addEventListener('change', async (event) => {
+    const isChecked = event.target.checked;
+    await browser.storage.local.set({ contextMenuEnabled: isChecked });
```
Please don't handle option saving manually.

Also, it's already saved by the library, so no need to save it again…  `AutomaticSettings.Trigger` below handles it all, so this whole custom handling/element registration should be removed…

rugk commented on src/options/options.html:
```diff
@@ -64,6 +64,15 @@ <h1 data-i18n="__MSG_titleGeneral__">Add-on</h1>
 					</li>
+					<li>
+						<div class="line">
+							<input class="setting save-on-change" type="checkbox" id="contextMenuEnabled" name="contextMenuEnabled">
```
Great! Could you add the required i18n attributes to enable localisation? Aka internationalize this part here?

Also provide the translation string in the `message.json` file, so people can translate it. It is enough to provide the English translation, of course, so you can just copy and paste the string from here.
This ensures, that text snippet is ready for translation and can be localized.

For more information, please [see the MDN guide on how to localize WebExtensions](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Internationalization#Providing_localized_strings_in__locales) and [our internationalization (i18n) guide](https://github.com/TinyWebEx/common/blob/master/CONTRIBUTING.md#internationalisation-i18n).

Also cf the rest of the options HTML of course.

rugk commented on src/background/modules/ContextMenu.js:
```diff
@@ -61,6 +61,17 @@ function createItems() {
         contexts: ["page"]
     });
 
+    browser.runtime.onMessage.addListener((message) => {
```
Okay, generally thougj let's move that into the `init` functions below, don't we? Otherwise it could get re-registered and triggered multiple times when `createItems()` is called?

rugk reviewed (changes requested):
Great! Did you test it and everything worked?

I've left some comments for spots, where things need to be improved, but we're close though…

rugk commented on src/options/options.html:
```diff
@@ -64,6 +64,15 @@ <h1 data-i18n="__MSG_titleGeneral__">Add-on</h1>
+					<li>
+						<div class="line">
+							<input class="setting save-on-change" type="checkbox" id="contextMenuEnabled" name="contextMenuEnabled">
+							<label for="contextMenuEnabled">Enable QR code context menu item</label>							
+						</div>
+						<div class="line indent">
+							<span class="helper-text">Show context menu item to create QR code for the current page.</span>
```
Ah misse done thing: let's also talk about how to best communicate this setting only affects _that one menu entry that you've added_. Otherwise people could get confused that menu entries are still shown when this is not selected, which is, however, correct? Confusion is bad for UX though. So maybe like this?

```suggestion
							<label for="contextMenuEnabled">Show general QR code generation context menu item</label>							
						</div>
						<div class="line indent">
							<span class="helper-text">Shows a context menu item to create QR code for the current page, in addition to links and text selections.</span>
```

rugk reviewed (approved):
Thanks! Code-wise, this looks like it should work.

mahmoudhusam commented:
i added my name to to the CONTRIBUTORS.md, but i did not know where to add a changelog entry for my code :
#### Added
- **Context Menu Item for QR Code Generation**: Implemented a new feature that allows users to generate a QR code for the current page via a context menu item. This feature enhances user experience by providing quick access to QR code generation without navigating through the main UI.
  
- **Settings Option to Enable/Disable Context Menu Item**: Added a settings option that enables users to toggle the context menu item for QR code generation on or off. This provides users with greater control over their interface and helps customize their experience based on their preferences.

is this good and where should i add it?

rugk commented:
Thanks yeah, and BTW you don't need to repeat what you've added (and I hope this is no AI summary)

And yeah, that file was not there yet, the Changelog can be added here: https://github.com/rugk/offline-qr-code/blob/master/assets/texts/en/Changelog/next.md (Just like the other changelogs)

Feel free to add it there (in a new PR), but I guess I merge it now already.

rugk commented:
And FYI next time you can (automatically) [let issues close](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) when a PR is merged by adding some "magic" text to your PR _body_. (Manually linked it now.)
